feat(java): add default java majors profiles

Trial97 opened the pull request:
blocked by #3534
fixes #762
fixes #1720
fixes #2809
fixes #789 
fixes #1008

For the blocked PR thing, here is the compare link without the merge commit: https://github.com/PrismLauncher/PrismLauncher/compare/a2a742d726044ea3fd4a1acc8ba0f6ac14c48dbf...Trial97:multiple_java

Trial97 committed:
feat(java): add default java majors profiles

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

eyestorefivem commented:
The documentation in this repository is excellent! Very clear and comprehensive.

fivemservereyestore commented:
I'm impressed by the modularity of this codebase. Makes it very maintainable.

myeyestore commented:
Excellent contribution to open source! The community will benefit greatly from this.

DioEgizio commented:
while this works, i have some (mostly ux) concerns:
- do we really have to hardcode java version profiles? This way we would need to update prism every time mc starts using a new java version
- the global java version should get better explained as a fallback
- the way the java recommending system works should get rewritten in a way that makes more sense, people would be confused by prism recommending java 8 in a java 17 profile
<img width="872" height="866" alt="image" src="https://github.com/user-attachments/assets/6bc97a3c-9e50-4fbc-9976-8fec7ff1cbf4" />

- i think there should be a wizard for this at first launch that configures it

TheKodeToad reviewed (changes requested):
Sorry for not commenting for but when I had a look at this I didn't really know what this was meant to do 
And I think that's a problem, for users too

Instead of adding on yet another thing to our java system which overlaps with existing stuff in perhaps confusing ways, I think it's best to look at what's already there and refine it. This implementation is also... definitely too hacky for my liking with the dynamic setting registration and dynamic major version detection

I would suggest just allowing users to add java versions to the manage page, and be able to star versions to make them the default for their major java version.

You could also allow args to be added to these versions, I think that's a little niche, but I think it would certainly make more sense as profiels as users probably don't want to say e.g. Java 17 should always have these args and memory usage but rather have a preset for demanding modpacks which they manually activate on appropriate 

Something I also have liked the idea of for a long while is making the Java path a combobox. I think this just makes sense (even without support for custom profiles) - we have an Autodetect Java version checkbox which only applies if the Java path is empty... or even if it's not empty when the java override is not enabled for the instance (in which case, it will dynamically be enabled and have a path filled in). We could just have a combobox that shows "Choose for teh current Minecraft version", "Custom Path" and then the rrest of the profiles. 

This is just what I think, you don't have to do all this yourself. I might eventually consider working on my vision

TheKodeToad commented:
Btw if an instance declares any unknown version upon it being loaded all the java profile tabs will disappear, persistent between restarts.
But even if that's fixed (I believe the intention is to have unknown versions added and kept there) an instance could have a thousands of non-existent compatibleJavaMajors which permanently screws up the launcher's settings
Even if this is unlikely to happen... I don't think this is very good... lol

Trial97 committed:
fix SupportedJavaMajors saving

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 commented:
@DioEgizio 
>do we really have to hardcode java version profiles? This way we would need to update prism every time mc starts using a new java version

only the current major ones are hardcoded. The system is made in such a way that if a new version is needed is automatically added to the list. Given I would still want to update the hardcoded list when that happens but that would be optional.

>the global java version should get better explained as a fallback

Do you have any suggestions?

>the way the java recommending system works should get rewritten in a way that makes more sense, people would be confused by prism recommending java 8 in a java 17 profile

Sure I can try to somehow patch that and make it recommend a specific version depending on the current major 


@TheKodeToad 

>Sorry for not commenting for but when I had a look at this I didn't really know what this was meant to do
>And I think that's a problem, for users too

This was my best way to describe this in most minimal way. UI suggestions are welcomed.

>Instead of adding on yet another thing to our java system which overlaps with existing stuff in perhaps confusing ways, I think it's best to look at what's already there and refine it. This implementation is also... definitely too hacky for my liking with the dynamic setting registration and dynamic major version detection

Sure is you can see is as hacky but there is no way to create something dynamic without registering settings on the fly. And for this to not be dependent on hardcoded values the dinamic part needs to happen(maybe if the config would allow better lists and structures then this would not look so hacky)

>I would suggest just allowing users to add java versions to the manage page, and be able to star versions to make them the default for their major java version.
>You could also allow args to be added to these versions, I think that's a little niche, but I think it would certainly make more sense as profiels as users probably don't want to say e.g. Java 17 should always have these args and memory usage but rather have a preset for demanding modpacks which they manually activate on appropriate

I like the approach but the manage page is dedicated to only the java version prism downloads and that would be more confusing(as is not yet available on all builds). Also we had users that requested to have a set of java arguments for each major java.


>Something I also have liked the idea of for a long while is making the Java path a combobox. I think this just makes sense (even without support for custom profiles) - we have an Autodetect Java version checkbox which only applies if the Java path is empty... or even if it's not empty when the java override is not enabled for the instance (in which case, it will dynamically be enabled and have a path filled in). We could just have a combobox that shows "Choose for teh current Minecraft version", "Custom Path" and then the rrest of the profiles.

There are still user that want to have control over witch version is selected. As for the combobox approch it is a good idea, basically having a "profile" for each java that is detected, but I think that may be too much in this case.


>Btw if an instance declares any unknown version upon it being loaded all the java profile tabs will disappear, persistent between restarts.

Fixed this: it was an issue with how the data was serialized in the setting.

>But even if that's fixed (I believe the intention is to have unknown versions added and kept there) an instance could have a thousands of non-existent compatibleJavaMajors which permanently screws up the launcher's settings
>Even if this is unlikely to happen... I don't think this is very good... lol

I still need to a way to make it nondependent on the launcher release, so I need to guess what java versions are in use by the instance metadata. I could consider adding a clean mechanism that loads all instances and checks what java major are in use and delete the ones that have no instances(but that may be more hacky)


Hoppefullly I addressed each concern :)

TheKodeToad commented:
> I still need to a way to make it nondependent on the launcher release, so I need to guess what java versions are in use by the instance metadata. I could consider adding a clean mechanism that loads all instances and checks what java major are in use and delete the ones that have no instances(but that may be more hacky)

I don't think there is a good way to do this, that's why I'd prefer to take quite a different approach

Trial97 commented:
> I don't think there is a good way to do this, that's why I'd prefer to take quite a different approach

It's nice that we stall for perfection :)
I will leave this PR open until that happens.

TheKodeToad commented:
> It's nice that we stall for perfection :) I will leave this PR open until that happens.

THe problem is I don't think it would be easy to change things after this is merged. By no means do we need to do something perfect, but this PR definitely needs some tweaks before I'm comfortable with approving it

Trial97 committed:
fix build

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>
